### ISSUE TITLE
[Feature] Psalm support

### ISSUE THREAD

**Issue by panosru** [php-censor/php-censor#276]
Are there any plans for [Psalm](https://getpsalm.org) support?

Useful resource:
https://medium.com/vimeo-engineering-blog/uncovering-php-bugs-with-template-a4ca46eb9aeb

Examples:
---
### Simulation of Generics

The following code is wrong because it defines the property `$bar` to be a type of `array<int>` which means an array with integer values only:
```php
<?php
declare(strict_types=1);

namespace Test;

class Foo
{
  /** @var array<int> */
  private $bar;

  /**
   * @param array<int> $bar
   */
  public function __construct(array $bar)
  {
    $this->bar = $bar;
  }
}


$foo = new Foo(['a']);
```

```
$ psalm
Scanning files...
Analyzing files...

ERROR: InvalidScalarArgument - src/test.php:21:16 - Argument 1 of Test\Foo::__construct expects array<array-key, int>, array{0:string(a)} provided
$foo = new Foo(['a']);


------------------------------
1 errors found
------------------------------

Checks took 0.22 seconds and used 23.997MB of memory
Psalm was able to infer types for 100.000% of analyzed code (1 file)
```

The following code is correct because it defines the `$bar` property to be a type of `array<string, int>` which means an array where key is string and value is integer:
```php
<?php
declare(strict_types=1);

namespace Test;

class Foo
{
  /** @var array<string, int> */
  private $bar;

  /**
   * @param array<string, int> $bar
   */
  public function __construct(array $bar)
  {
    $this->bar = $bar;
  }
}


$foo = new Foo(['a' => 1]);
```

```
$ psalm
Scanning files...
Analyzing files...

------------------------------
No errors found!
------------------------------

Checks took 0.26 seconds and used 23.779MB of memory
Psalm was able to infer types for 100.000% of analyzed code (1 file)
```

The following code is correct because it defines the `$bar` property to be a type of `array<string, Bar>` which means an array where key is string and value is instance of `Bar` object:
```php
<?php
declare(strict_types=1);

namespace Test;

class Foo
{
  /** @var array<string, Bar> */
  private $bar;

  /**
   * @param array<string, Bar> $bar
   */
  public function __construct(array $bar)
  {
    $this->bar = $bar;
  }
}

class Bar {}

$foo = new Foo(['a' => new Bar()]);
```

```
$ psalm
Scanning files...
Analyzing files...

------------------------------
No errors found!
------------------------------

Checks took 0.28 seconds and used 23.810MB of memory
Psalm was able to infer types for 100.000% of analyzed code (1 file)
```

While that is the primary function I use, [there is plenty of functionality](https://getpsalm.org/docs/supported_annotations) that is worth checking.

**Comment by corpsee** [comment]
@panosru Yes, but it is not high priority feature.

**Comment by panosru** [comment]
@corpsee I'm not saying that it is :) It's just a great tool to have, I personally use it in all my projects. 

So far I have created a Psalm plugin in my local php-censor testing environment and I'm playing around with the code.

I had trouble installing psalm as a local dependency because of compatibility issues with other plugins, so I'm using the global version for now.

From what I understood from other plugins you have, you basically need to catch errors or successes with regular expressions and pass that to php-censor so for it to understand if the plugin failed or passed the test, correct?

So the basic strategy is to run psalm binary and read the output, from that ouput catch errors and warning or success, am I missing something?

Thanks!

**Comment by corpsee** [comment]
@panosru If you create plugin it wil be awesome!

* About dependencies. I think you are about minimal version of PHP (`^7.0`). You can write to plugin documentation that plugin works only on PHP `7.0` installation and that user should have `system` or `local` Psalm and that option `global` (PHP Censor dependencies) are not supported.

    By the way PHP Censor version `2.0` will have minimal PHP version `7.1` in the future.

* About implementation. You can implement simple plugin with minimum functionality: execute Psalm and write output log to PHP Censor log (See [Atoum plugin code](https://github.com/php-censor/php-censor/blob/master/src/Plugin/Atoum.php) for example). Or you can implement more complex plugin with parsing/storing build errors and build meta, presentation additional js charts, tables etc.

**Comment by panosru** [comment]
@corpsee could you please tell me how to invoke a build without having to run workers and etc, in order to make my dev easier?

**Comment by corpsee** [comment]
@panosru No way. Only by worker or cronjob.

**Comment by garas** [comment]
>felixfbecker/advanced-json-rpc v3.0.3 requires phpdocumentor/reflection-docblock ^4.0.0
> php-censor/php-censor dev-master requires phpdocumentor/reflection-docblock ~2.0.0

But I didn't find any `phpDocumentor\Reflection` within code. PhpDocblockChecker plugin uses `php-censor/phpdoc-checker` for its job.

@corpsee It probably left after switch to own until. Remove this dependency. Less dependencies, less problems.

**Comment by panosru** [comment]
So far my plugin looks like this:
```php
<?php

namespace PHPCensor\Plugin;

use PHPCensor;
use PHPCensor\Builder;
use PHPCensor\Model\Build;
use PHPCensor\Model\BuildError;
use PHPCensor\Plugin;
use PHPCensor\ZeroConfigPluginInterface;

/**
 * A static analysis tool for finding errors in PHP applications https://getpsalm.org
 *
 * Class Psalm
 * @package PHPCensor\Plugin
 */
class Psalm extends Plugin implements ZeroConfigPluginInterface
{
    /**
     * @return string
     */
    public static function pluginName()
    {
        return 'psalm';
    }

    /**
     * {@inheritdoc}
     */
    public static function canExecuteOnStage($stage, Build $build)
    {
        return Build::STAGE_TEST === $stage;
    }

    /**
     * {@inheritdoc}
     */
    public function __construct(Builder $builder, Build $build, array $options = [])
    {
        parent::__construct($builder, $build, $options);

        $this->executable = $this->findBinary('psalm');
    }

    /**
     * Runs Psalm
     */
    public function execute()
    {
        $psalm = $this->executable;

        $this->builder->executeCommand('cd "%s" && ' . $psalm . ' --output-format=json', $this->builder->buildPath);
        $output  = \json_decode($this->builder->getLastOutput());

        // Define that the plugin succeed
        $success = true;

        if (null !== $output) {
            $errors = [];
            $infos = [];

            if (\is_array($output) && 0 < \count($output)) {
                foreach ($output as $value) {
                    if (!\in_array($value->severity, ['error','info'])) {
                        continue;
                    }

                    // Template:
                    // MissingConstructor - src/Plugins/Foobar.php:29:15 - Foo\Bar has an uninitialized variable $this->filesystem, but no constructor
                    \array_push(${$value->severity.'s'}, [
                        'full_message' => \vsprintf('%s - %s:%d:%d - %s' . PHP_EOL . '%s', [
                            $value->type,
                            $value->file_name,
                            $value->line_from,
                            $value->column_from,
                            $value->message,
                            $value->snippet
                        ]),
                        'message' => $value->message,
                        'file' => $value->file_name,
                        'line_from' => $value->line_from,
                        'line_to' => $value->line_to
                    ]);
                }
            }

            if (0 < \count($errors)) {
                foreach ($errors as $error) {
                    $this->builder->logFailure($error['full_message']);

                    $this->build->reportError(
                        $this->builder, self::pluginName(), $error['message'], BuildError::SEVERITY_HIGH, $error['file'], $error['line_from'], $error['line_to']
                    );
                }
            }

            if (0 < \count($infos)) {
                foreach ($infos as $info) {
                    if (true === true) {
                        $this->builder->logFailure($info['full_message']);
                    }

                    $this->build->reportError(
                        $this->builder, self::pluginName(), $info['message'], BuildError::SEVERITY_LOW, $info['file'], $info['line_from'], $info['line_to']
                    );
                }
            }

        } else {
            $success = false;
            $this->builder->logFailure('No tests have been performed.');
        }

        if ($success) {
            $this->builder->logSuccess('No errors found!');
        }


        return $success;
    }
}
```

I'm still working on it, I have some questions though.

In my build log tab I would like to hide the json output and only send custom output to the user, instead I'm getting this:
![image](https://user-images.githubusercontent.com/400362/52664150-f5b87700-2f10-11e9-83bf-b7d4bca84b14.png)

While in my code I'm using `$this->builder->logFailure` and `$this->build->reportError` for some reason it does not seem to work.

In order for me to test easily my project I have created a test php file with the following code in the root of the repository I'm testing in my local env:

```php
<?php
declare(strict_types=1);

$output = shell_exec('psalm  --output-format=json');

$output = json_decode($output);

if (null !== $output) {
  $errors = [];
  $infos = [];

  if (\is_array($output) && 0 < \count($output)) {
    foreach ($output as $value) {
      if (!\in_array($value->severity, ['error','info'])) {
        continue;
      }

      // Template:
      // MissingConstructor - src/Plugins/Foobar.php:29:15 - Foo\Bar has an uninitialized variable $this->filesystem, but no constructor
      \array_push(${$value->severity.'s'}, \vsprintf('%s - %s:%d:%d - %s' . PHP_EOL . '%s', [
        $value->type,
        $value->file_name,
        $value->line_from,
        $value->column_from,
        $value->message,
        $value->snippet
      ]));
    }
  }

  echo 'Total issues: ' . \count($output) . PHP_EOL;
  echo 'Info count: ' . \count($infos) . PHP_EOL;
  echo 'Info Example: ' . PHP_EOL . $infos[0] . PHP_EOL;
  echo 'Error count: ' . \count($errors) . PHP_EOL;
  echo 'Error Example: ' . PHP_EOL . $errors[0];

} else {
  echo "Is null";
}
```

and the output:
![image](https://user-images.githubusercontent.com/400362/52664450-bdfdff00-2f11-11e9-8356-493107eff674.png)

Any hint would be much appriciated!

**Comment by garas** [comment]
For output control, look how `PhpCodeSniffer` plugin handles this

**Comment by panosru** [comment]
@garas Ah! `$this->builder->logExecOutput` nice!

I will play with it! If I have any more questions I'll post here!

Thanks!

**Comment by panosru** [comment]
@garas It seems that I have also to restart my worker every time I make any change to my plugin :/ :/ 

Nevertheless, I have some progress:
![image](https://user-images.githubusercontent.com/400362/52666297-93fb0b80-2f16-11e9-9654-7fb4a22aa752.png)

![image](https://user-images.githubusercontent.com/400362/52666327-ab39f900-2f16-11e9-86a0-cbbfa479e7ac.png)


I'm trying to figure out how to add data to `Information` tab as well.

Once I have a clean code and the desired output result, I'll make a PR

**Comment by panosru** [comment]
Ok, I pretty much got it working, I have a few questions:

1. In case the user set `allowed_errors` to `1000` and accordingly `allowed_warnings`, should I record the errors and warnings in the error tab or should I skip them and record only those that exceed the number that the user difined?

2. Psalm has `errors` and `infos`, I'm using `allowed_warnings` settings for `infos` to keep the consistency of the php-censor and being able to inherit from global settings, what do you think about that? Is that a good way or will cause confusion to the user? *(of course I plan to add into the docs that the `allowed_warnings` setting is refered to `infos`)*

3. What exactly the `$this->build->storeMeta((self::pluginName() . '-warnings'), $infos);` and `$this->build->storeMeta((self::pluginName() . '-errors'), $errors);` do? How and where can I use this meta data?

4. Is the output acceptable?
![image](https://user-images.githubusercontent.com/400362/52673794-64ee9500-2f2a-11e9-9ef6-4bd7307b628f.png)
![image](https://user-images.githubusercontent.com/400362/52673828-79329200-2f2a-11e9-91f9-e1c384946d73.png)
![image](https://user-images.githubusercontent.com/400362/52673858-8c456200-2f2a-11e9-8d90-16375a8578cd.png)


Thanks!

**Comment by corpsee** [comment]
@panosru 

1. You should record errors anyway.
2. I don't know. I guess it is question for PR review :)
3. It is meta information for js charts and etc. (Quality Trend).
4. You should use PHP version 5.6 if you want to add plugin to main repository. If you create plugin as separate repository you may use any PHP version.
5. Sure.

**Comment by corpsee** [comment]
@garas I will check dependencies soon. Thanks.

**Comment by panosru** [comment]
@corpsee Thanks for the answers!

For the php code, I'll do some refactoring and it will support 5.6, not a big deal.

For the quality trend, could you point me to a plugin(s) where I could get more info?

Or on any information where I could read a bit more about it so I could make use of it...

Thanks!

PS: Now that I got the main point on how to make plugins, I will re-check [that list](https://github.com/exakat/php-static-analysis-tools) to see maybe I'll make use of something in addition to Psalm :)

**Comment by corpsee** [comment]
@panosru About quality trend [see "warnings.js"](https://github.com/php-censor/php-censor/blob/master/public/assets/js/build-plugins/warnings.js). Also you can create personal plugin's widget like [PHPLoc](https://github.com/php-censor/php-censor/blob/master/public/assets/js/build-plugins/loc.js), [PHPUnit](https://github.com/php-censor/php-censor/blob/master/public/assets/js/build-plugins/phpunit.js) or [Codeception](https://github.com/php-censor/php-censor/blob/master/public/assets/js/build-plugins/codeception.js) (Optional and unique).

**Comment by garas** [comment]
@corpsee I've checked again about dependencies.

`vimeo/psalm` install is blocked by `phpdocumentor/reflection-docblock`, but as I've said, I didn't found any usage.

Another blocker is `php-censor/phpdoc-checker 1.0.3 (requires nikic/php-parser 3.*)` while `vimeo/psalm 3.0.0 requires nikic/php-parser ^4.0.2 || ^4.1`. Again, it is own package, might be possibly upgraded.  

After these, `composer require vimeo/psalm` succeeds.

**Comment by corpsee** [comment]
@garas Thank you! Could you create PR for this problem please (If you have possibility)?

**Comment by panosru** [comment]
@corpsee I will try to solve the composer issue today, once I got it right I will push the commit to my PR.

I also plan to create plugins for the following tools:

- [Phlint](https://gitlab.com/phlint/phlint) - A tool with an aim to help maintain the quality of PHP code by analyzing code and pointing out potential code issues.
- [PHPStan](https://github.com/phpstan/phpstan) - Focuses on finding errors in code without actually running it.
- [Pahout](https://github.com/wata727/pahout) - A pair programming partner for writing better PHP.

In order for me to push new PRs, should I first wait for my current PR be merged? Put everything in the current PR (bad practice AFAIK)? or create new branches in my fork and make isolated PRs for each new plugin (and additional changes in composer)?

Thanks!

**Comment by corpsee** [comment]
@panosru I think the best way is create isolated new branches/PR. Thanks.

**Comment by garas** [comment]
I forgot Psalm requires PHP 7, so `composer require vimeo/psalm` will not work on PHP 5.6. Same for `nikic/php-parser ^4.0`.

1.x needs `composer global require vimeo/psalm`, and local will need to go 2.x branch anyway. I have not enough time now to upgrade `php-censor/phpdoc-checker` to support `nikic/php-parser ^4.0`, maybe later, 2.x just started, so no hurry yet anyway.

**Comment by corpsee** [comment]
The PR (#277) merged to `master` branch and will be released with next minor version `1.1.0`.
